packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,268 @@
+import { M, matches } from '@endo/patterns';
+import { AmountShape, BrandShape } from '@agoric/ertp';
+import { makeICAConnectionAddress, parseAddress } from '@agoric/orchestration';
+import '@agoric/network/exported.js';
+
+const { Fail, bare } = assert;
+const trace = makeTracer('Orchestration');
+
+// XXX improve me
+/** @typedef {string} ChainAddress */
+
+/**
+ * @typedef {object} OrchestrationPowers
+ * @property {ERef<
+ *   import('@agoric/orchestration/src/types').AttenuatedNetwork
+ * >} network
+ */
+
+/**
+ * PowerStore is used so additional powers can be added on upgrade.
+ *
+ * @typedef {MapStore<
+ *   keyof OrchestrationPowers,
+ *   OrchestrationPowers[keyof OrchestrationPowers]
+ * >} PowerStore
+ */
+
+/**
+ * @template {keyof OrchestrationPowers} K
+ * @param {PowerStore} powers
+ * @param {K} name
+ */
+const getPower = (powers, name) => {
+  powers.has(name) || Fail`need powers.${bare(name)} for this method`;
+  return /** @type {OrchestrationPowers[K]} */ (powers.get(name));
+};
+
+export const ChainAccountI = M.interface('ChainAccount', {
+  getAccountAddress: M.call().returns(M.string()),
+  getLocalAddress: M.call().returns(M.string()),
+  getRemoteAddress: M.call().returns(M.string()),
+  getPort: M.call().returns(M.remotable('Port')),
+  executeTx: M.callWhen(M.arrayOf(M.record())).returns(M.any()),
+  executeEncodedTx: M.callWhen(M.string()).returns(M.any()),
+  deposit: M.callWhen(M.remotable('Payment'))
+    .optional(M.pattern())
+    .returns(AmountShape),
+  getPurse: M.callWhen(BrandShape).returns(M.remotable('VirtualPurse')),
+  close: M.callWhen().returns(M.promise()),
+});
+
+export const ConnectionHandlerI = M.interface('ConnectionHandler', {
+  onOpen: M.callWhen(M.any(), M.string(), M.string(), M.any()).returns(M.any()),
+  onClose: M.callWhen(M.any(), M.string()).returns(M.any()),
+  onReceive: M.callWhen(M.any(), M.string()).returns(M.any()),
+});
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareChainAccount = zone =>
+  zone.exoClassKit(
+    'ChainAccount',
+    { account: ChainAccountI, connectionHandler: ConnectionHandlerI },
+    /**
+     * @param {Port} port
+     * @param {string} requestedRemoteAddress
+     */
+    (port, requestedRemoteAddress) =>
+      /**
+       * @type {{
+       *   port: Port;
+       *   connection: Connection | undefined;
+       *   localAddress: string | undefined;
+       *   requestedRemoteAddress: string;
+       *   remoteAddress: string | undefined;
+       *   accountAddress: ChainAddress | undefined;
+       * }}
+       */ (
+        harden({
+          port,
+          connection: undefined,
+          requestedRemoteAddress,
+          remoteAddress: undefined,
+          accountAddress: undefined,
+          localAddress: undefined,
+        })
+      ),
+    {
+      account: {
+        getAccountAddress() {
+          const { accountAddress } = this.state;
+          accountAddress || Fail`account address not available`;
```

Comment:
any advice for the caller?

Reply:
Updated to `Error parsing account address from remote address` in 2a39e6dc